components/net/local_directory_listing.rs

```diff
@@ -0,0 +1,388 @@
+use net_traits::{NetworkError, ResourceFetchTiming};
+use servo_config::pref;
+use servo_url::ServoUrl;
+use time_03::OffsetDateTime;
+use url::Url;
+
+pub fn is_request_allowed(request: &Request) -> Result<(), Response> {
+    if !pref!(network.local_directory_listing.enabled) {
+        // If you want to be able to browse local directories, configure Servo prefs so that
+        // "network.local_directory_listing.enabled" is set to true.
+        Err(Response::network_error(NetworkError::Internal(
+            "Local directory listing feature has not been enabled in preferences".into(),
+        )))
+    } else if !request.origin.is_opaque() {
+        // Checking for an opaque origin as a shorthand for user activation
+        // as opposed to a request originating from a script.
+        // TODO: Raise and link to GitHub issue to request consideration of this concern.
+        Err(Response::network_error(NetworkError::Internal(
+            "Cannot request local directory listing from non-local origin.".to_string(),
+        )))
+    } else {
+        Ok(())
+    }
+}
+
+pub fn fetch(
+    request: &mut Request,
+    url: ServoUrl,
+    directory: impl Into<DirectorySummary>,
+) -> Response {
+    let url = if !url.path().ends_with('/') {
+        // Re-read the path as a directory, so that Servo adds the
+        // forward-slash to the end of the URL (at least internally,
+        // though the URL bar does not reflect this) and loads the
+        // linked directories and files correctly when clicked.
+        if let Ok(dir_url) = Url::from_directory_path(url.path()) {
+            ServoUrl::from_url(dir_url)
+        } else {
+            return Response::network_error(NetworkError::Internal(format!(
+                "Unable to parse local directory path {}",
+                url
+            )));
+        }
+    } else {
+        url
+    };
+    let mut response = Response::new(url, ResourceFetchTiming::new(request.timing_type()));
+    response.headers.typed_insert(ContentType::html());
+    let directory_summary = directory.into();
+    let page_text = build_html_directory_listing(directory_summary);
+    let bytes: Vec<u8> = page_text.into_bytes();
+    *response.body.lock().unwrap() = ResponseBody::Done(bytes);
+    response
+}
+
+impl Into<DirectorySummary> for PathBuf {
+    fn into(self) -> DirectorySummary {
+        let path = self
+            .to_str()
+            .map(str::to_string)
+            .ok_or("Invalid directory path.");
+        let has_parent = self.parent().is_some();
+        let items = if let Ok(entries) = std::fs::read_dir(self) {
+            Ok(gather_directory_items(entries))
+        } else {
+            Err("Unable to iterate directory contents.")
+        };
+        DirectorySummary {
+            path,
+            has_parent,
+            items,
+        }
+    }
+}
+
+fn gather_directory_items(
+    entries: std::fs::ReadDir,
+) -> BTreeMap<OsString, DirectoryItemDescriptor> {
```

Comment:
The `BTreeMap` is being used to keep the directory listing in alphabetical order (sorted by the value of the `OsString` of each file/symlink/directory name in the listing).

`(OsString, Vec<_>)` isn't the same structure, so I don't follow how this could replace the `BTreeMap` here. If raw efficiency was critical, then it might be better to simply hold a `Vec<DirectoryItemDescriptor>` while building up the listing, then sort just once before rendering the HTML. But in a method triggered only once every few seconds by user interaction, that sort of small efficiency saving is usually unnecessary.